### PR TITLE
gh-96711: Do not print debug information upon Invalid opcode

### PR DESCRIPTION
**Pull request by serge-sans-paille**
* Issue: gh-96711

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
cc @vstinner :-)

File: Python/ceval.c [hunk 1]
```diff
@@ -5059,8 +5059,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             /* Tell C compilers not to hold the opcode variable in the loop.
                next_instr points the current instruction without TARGET(). */
             opcode = _Py_OPCODE(*next_instr);
+#ifndef NDEBUG
```
Comment (vstinner) [inline Python/ceval.c]: I don't see the point of polluting stderr if the caller is able to catch SystemError. I suggest removing the printf, but please complete the SystemError error message to add the info which are currently available in the printf (opcode, line number, etc.). Maybe including the code name / function name?

**Comment by serge-sans-paille** [comment]
@vstinner : as green as a clover :-)

File: Lib/test/test_code.py [hunk 2]
```diff
@@ -337,6 +337,37 @@ def func():
         new_code = code = func.__code__.replace(co_linetable=b'')
         self.assertEqual(list(new_code.co_lines()), [])
 
+    def test_invalid_bytecode(self):
+        def foo(): pass
+        co = foo.__code__
+        CodeType = type(co)
+
+        new_co = CodeType(
```
Comment (vstinner) [inline Lib/test/test_code.py]: Please don't use the CodeType constructor, it's a pain to maintain. I would suggest to use foo.__code__.replace(co_code=...) instead. Maybe something like:

```
code = b'\xEE' + foo.__code__
foo.__code__.replace(co_code=code)
```

File: Python/ceval.c [hunk 3]
```diff
@@ -5051,9 +5051,11 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             /* Tell C compilers not to hold the opcode variable in the loop.
                next_instr points the current instruction without TARGET(). */
             opcode = _Py_OPCODE(*next_instr);
-            fprintf(stderr, "XXX lineno: %d, opcode: %d\n",
-                    _PyInterpreterFrame_GetLine(frame),  opcode);
-            _PyErr_SetString(tstate, PyExc_SystemError, "unknown opcode");
+            _PyErr_Format(tstate, PyExc_SystemError,
+                          "%U:%d: unknown opcode '%d'",
```
Comment (vstinner) [inline Python/ceval.c]:
```suggestion
                          "%U:%d: unknown opcode %d",
```

File: Lib/test/test_code.py [hunk 4]
```diff
@@ -337,6 +337,37 @@ def func():
...
+            co.co_firstlineno,
+            co.co_linetable,
+            co.co_exceptiontable,
+            co.co_freevars,
+            co.co_cellvars)
+
+        foo.__code__ = new_co
+        with self.assertRaises(SystemError) as se:
+            foo()
+        self.assertIn("unknown opcode '238'", str(se.exception))
```
Comment (vstinner) [inline Lib/test/test_code.py]: Would it be possible to check the filename, and maybe even the line number? Try values from ``foo.__code__``?

* ``foo.__code__.co_filename``
* ``foo.__code__.co_firstlineno``

File: Lib/test/test_code.py (hunk 2 above)
Comment (serge-sans-paille) [reply Lib/test/test_code.py]: I wasn't aware of this method, that's indeed much more pleasant to the eye!

**Comment by vstinner** [comment]
Merged, thanks @serge-sans-paille. It's good that all data are now available to the caller in the exception message directly, it's more convenient than having to deal with stderr.

IMO it's an enhancement, not a bugfix, so I will not backport the change to Python 3.11 and older versions.

**Comment by serge-sans-paille** [comment]
Thanks for the review!

**Pull request merged**